src/Raven.Server/Commercial/SetupManager.cs

```diff
@@ -0,0 +1,365 @@
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException("Failed to complete dns challenge from Let's Encrypt.", e);
+            }
+            return progress;
+        }
+
+        public static string WriteSettingsJsonFile(string base64Cert, string publicUrl, string serverUrl, string settingsPath, SetupMode setupMode, bool modifyLocalServer)
+        {
+            var json = File.ReadAllText(settingsPath);
+            dynamic jsonObj = JsonConvert.DeserializeObject(json);
+
+            jsonObj["ServerUrl"] = serverUrl;
+            jsonObj["PublicServerUrl"] = publicUrl;
+            jsonObj["Security.Certificate.Base64"] = base64Cert;
+            jsonObj["Setup.Mode"] = setupMode.ToString();
+
+            string output = JsonConvert.SerializeObject(jsonObj, Formatting.Indented);
+
+            if (modifyLocalServer)
+            {
+                var tmpPath = settingsPath + ".tmp";
+                using (var file = new FileStream(tmpPath, FileMode.Create))
+                using (var writer = new StreamWriter(file))
+                {
+                    writer.Write(output);
+                    writer.Flush();
+                    file.Flush(true);
+                }
+
+                File.Replace(tmpPath, settingsPath, settingsPath + ".bak");
+                if (PlatformDetails.RunningOnPosix)
+                    Syscall.FsyncDirectoryFor(settingsPath);
+            }
+
+            return output;
+        }
+
+        private class UniqueResponseResponder : IStartup
+        {
+            private readonly string _response;
+
+            public UniqueResponseResponder(string response)
+            {
+                _response = response;
+            }
+
+            public IServiceProvider ConfigureServices(IServiceCollection services)
+            {
+                return services.BuildServiceProvider();
+            }
+
+            public void Configure(IApplicationBuilder app)
+            {
+                app.Run(async context =>
+                {
+                    await context.Response.WriteAsync(_response);
+                });
+            }
+        }
+
+        public async void AssertServerCanStartSecured(X509Certificate2 serverCertificate, string serverUrl, IPEndPoint[] addresses, string settingsPath)
+        {
+            var configuration = new RavenConfiguration(null, ResourceType.Server, settingsPath);
+            configuration.Initialize();
+
+            var guid = Guid.NewGuid().ToString();
+            var responder = new UniqueResponseResponder(guid);
+
+            var webHost = new WebHostBuilder()
+                .CaptureStartupErrors(captureStartupErrors: true)
+                .UseKestrel(options =>
+                {
+                    if (addresses.Length == 0)
+                        options.Listen(new IPEndPoint(IPAddress.Parse("0.0.0.0"), 8080), listenOptions => listenOptions.UseHttps(serverCertificate));
```

Comment:
`8080` as the port is not likely to be what we want. We will usually want to get either `443`, if the user didn't specify it or have the user explicitly select it.